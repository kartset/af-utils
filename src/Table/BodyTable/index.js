import React from "react";
import { observer } from "mobx-react-lite";
import useApi from "../../useApi";
import cx from "../../utils/cx";
import css from "./style.module.scss";

const BodyTable = ({ children, className }) => {

    const API = useApi();

    return (
        <table
            children={children}
            className={cx(css.wrapper,className)}
            aria-rowcount={API.Rows.visibleRowCount}
            aria-colcount={API.columns.length}
        />
    );
};

export default observer( BodyTable );