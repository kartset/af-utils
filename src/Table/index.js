import React, { useLayoutEffect, memo } from "react";
import PropTypes from "prop-types";

import isPositionStickySupported from "../utils/isPositionStickySupported";
import Context from "../Context";
import useStore from "../utils/useStore";

import FixedSizeTableStore from "../models/FixedSizeTable";
import VariableSizeTableStore from "../models/VariableSizeTable";

import RowComponentDefault from "./common/Row";
import CellComponentDefault from "./common/Cell";
import TotalsCellComponentDefault from "./common/TotalsCell";

import NonStickyComponent from "./NonSticky";
import StickyComponent from "./Sticky";

import commonPropTypes from "../commonPropTypes";
import commonDefaultProps from "../commonDefaultProps";
import cx from "../utils/cx";

import castArray from "lodash/castArray";

const Table = ({
    fixedSize,
    estimatedRowHeight,
    columns,
    totals,
    getRowData,
    getCellData,
    getRowKey,
    getRowExtraProps,
    getCellExtraProps,
    rows,
    overscanRowsCount,
    headless,
    dataRef,
    nonSticky,
    className,
    initialGrouping,
    initialExpandedGroups,
    ...props
}) => {

    const [ Store, scrollContainerRef, tbodyRef ] = useStore( fixedSize ? FixedSizeTableStore : VariableSizeTableStore, dataRef, {
        headlessMode: headless,
        getRowData,
        getCellData,
        getRowKey,
        overscanRowsCount,
        estimatedRowHeight,
        totals,
        columns,
        rows
    });

    useLayoutEffect(() => {
        if( initialGrouping ){
            Store.Rows.aggregators.addGrouping( ...castArray( initialGrouping ) );
            if( initialExpandedGroups ){
                Store.Rows.resetExpandedState( initialExpandedGroups )
            }
        }
    }, [ Store ]);

    /*
        Only cells inside thead/tfoot can be sticky.
        If thead/tfoot are hidden - we can easily render lighter StickyComponent to avoid extra wrappers
    */
    const ComponentVariant = ( headless && !totals ) || ( !nonSticky && isPositionStickySupported() ) ? StickyComponent : NonStickyComponent;

    return (
        <Context.Provider value={Store}>
            <ComponentVariant
                className={cx("afvscr-table-wrapper",className)}
                scrollContainerRef={scrollContainerRef}
                getRowExtraProps={getRowExtraProps}
                getCellExtraProps={getCellExtraProps}
                tbodyRef={tbodyRef}
                {...props}
            />
        </Context.Provider>
    );
}

Table.propTypes = {
    ...commonPropTypes,
    columns: PropTypes.arrayOf(
        PropTypes.shape({
            // unique key for column
            dataKey: PropTypes.string.isRequired,

            // for details see CellComponent implementation
            getCellData: PropTypes.func,
            getEmptyCellData: PropTypes.func,
            format: PropTypes.func,
            render: PropTypes.func,
            formatTotal: PropTypes.func,

            visibility: PropTypes.oneOf([ "visible", "hidden" ]),
            sort: PropTypes.oneOf([ "locale", "numeric" ]),

            // column props, affecting colgroup > col tags
            background: PropTypes.string,
            border: PropTypes.string,
            width: PropTypes.oneOfType([ PropTypes.number, PropTypes.string ]),
            CellComponent: PropTypes.elementType,
            getCellExtraProps: PropTypes.func
        })
    ).isRequired,

    getCellExtraProps: PropTypes.func,

    totals: PropTypes.objectOf(
        // array may contain: "sum", "average", "count", "max", "min"
        PropTypes.array
    ),
    
    nonSticky: PropTypes.bool,
    headless: PropTypes.bool,

    HeaderRowComponent: PropTypes.elementType,
    CellComponent: PropTypes.elementType,
    getCellData: PropTypes.func,
    TotalsCellComponent: PropTypes.elementType,

    initialGrouping: PropTypes.oneOfType([ PropTypes.string, PropTypes.array ]),
    initialExpandedGroups: PropTypes.object
};

Table.defaultProps = {
    ...commonDefaultProps,
    headless: false,

    //    For 90% non-reactive solutions, which only provide new getRowData when data is changed, memo is ok.
    //    If RowComponent should be wrapped my mobx observer - non-memo version should be imported.
    //    memo(observer(RowComponentDefault)) will do the trick.
    
    RowComponent: memo( RowComponentDefault ),
    CellComponent: CellComponentDefault,
    getCellData: ( rowData, rowIndex, dataKey ) => rowData[ dataKey ],
    TotalsCellComponent: TotalsCellComponentDefault
};

export default memo( Table );