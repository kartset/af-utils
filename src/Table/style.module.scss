
.wrapper {

    thead,
    tfoot {
        white-space: nowrap;
    }

    td,
    th {
        overflow: hidden;
        text-overflow: ellipsis;
    }

    th[data-sortable]{
        cursor: pointer;
        user-select: none;
    }
    
    th[aria-sort="ascending"]::after{
        content: " ↑";
    }
    
    th[aria-sort="descending"]::after{
        content: " ↓";
    }

    thead td,
    thead th,
    tfoot td,
    tfoot th {
        position: sticky;
        z-index: 1;
    }

    thead td,
    thead th {
        top: 0;
    }
}