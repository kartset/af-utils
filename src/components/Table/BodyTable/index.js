import { memo } from "react";
import useModelSubscription from "hooks/useModelSubscription";
import cx from "utils/cx";
import css from "./style.module.scss";

const BODY_TABLE_SUBSCRIPTIONS = [ "rowsQuantity", "normalizedVisibleColumns" ];

const BodyTable = ({ children, className }) => {

    const { rowsQuantity, normalizedVisibleColumns } = useModelSubscription( BODY_TABLE_SUBSCRIPTIONS );

    return (
        <table
            children={children}
            className={cx(css.wrapper,className)}
            aria-rowcount={rowsQuantity}
            aria-colcount={normalizedVisibleColumns.length}
        />
    );
};

export default memo( BodyTable );