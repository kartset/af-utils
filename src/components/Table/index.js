import { memo } from "react";
import PropTypes from "prop-types";

import commonPropTypes from "../common/propTypes";
import cx from "utils/cx";

import Context from "Context";
import useModel from "hooks/useModel";

import FixedTable from "models/tables/SimpleFixed";
import VariableTable from "models/tables/SimpleVariable";

import ScrollContainer from "../common/ScrollContainer";
import Scroller from "../common/Scroller";

import Rows from "./Rows";
import Colgroup from "./Colgroup";
import HeaderCells from "./HeaderCells";

import {
    renderRow,
    renderCell,
    CellsList,
    Cell,
    HeaderCell
} from "./renderers";

import css from "./style.module.scss";

/*
    Todo:
        * measure thead & tfoot heights in order to properly calculate available space for rows
        * think about border-collapse offsetHeight issue
        * maybe throw border-collapse
*/


const Table = ({
    fixed,
    estimatedRowHeight,
    columns,
    getRowData,
    renderRow,
    renderCell,
    CellsList,
    Cell,
    HeaderCell,
    rowsQuantity,
    overscanRowsCount,
    headless,
    dataRef,
    className,
    onHeaderCellClick,
    ...props
}) => {

    const [ Store, rowsContainerRef ] = useModel( fixed ? FixedTable : VariableTable, dataRef, {
        overscanRowsCount,
        estimatedRowHeight,
        columns,
        rowsQuantity
    });
    
    return (
        <Context.Provider value={Store}>
            <ScrollContainer className={cx(css.wrapper,className)} {...props}>
                <table className={css.bodyTable}>
                    <Colgroup />
                    {headless?null:(
                        <thead>
                            <tr>
                                <HeaderCells onClick={onHeaderCellClick} HeaderCell={HeaderCell} />
                            </tr>
                        </thead>
                    )}
                    <Scroller as={<tbody />} />
                    <tbody ref={rowsContainerRef}>
                        <Rows
                            getRowData={getRowData}
                            renderRow={renderRow}
                            renderCell={renderCell}
                            CellsList={CellsList}
                            Cell={Cell}
                        />
                    </tbody>
                </table>
            </ScrollContainer>
        </Context.Provider>
    );
}

Table.propTypes = {
    ...commonPropTypes,
    columns: PropTypes.arrayOf(
        PropTypes.oneOfType([
            PropTypes.string,
                PropTypes.shape({
                // unique key for column
                dataKey: PropTypes.string.isRequired,

                // for details see CellComponent implementation
                getEmptyCellData: PropTypes.func,
                format: PropTypes.func,
                render: PropTypes.func,
                formatTotal: PropTypes.func,

                visibility: PropTypes.oneOf([ "visible", "hidden" ]),

                // column props, affecting colgroup > col tags
                background: PropTypes.string,
                border: PropTypes.string,
                width: PropTypes.oneOfType([ PropTypes.number, PropTypes.string ]),
                CellComponent: PropTypes.elementType,
                getCellExtraProps: PropTypes.func
            })
        ])
    ).isRequired,

    getRowData: PropTypes.func.isRequired,

    renderRow: PropTypes.func,
    renderCell: PropTypes.func,
    CellsList: PropTypes.elementType,
    Cell: PropTypes.elementType,
    HeaderCell: PropTypes.elementType,

    headless: PropTypes.bool,
};

Table.defaultProps = {
    headless: false,
    fixed: false,
    estimatedRowHeight: 20,
    overscanRowsCount: 4,

    renderRow,
    renderCell,
    CellsList,
    Cell,
    HeaderCell
};

export default memo( Table );