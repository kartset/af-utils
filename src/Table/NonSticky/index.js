import React from "react";
import { observer } from "mobx-react-lite";
import PropTypes from "prop-types";

import useApi from "../../useApi";
import TableWrapper from "./TableWrapper";

import Thead from "../common/Thead";
import Tfoot from "../common/Tfoot";
import Tbody from "../common/Tbody";
import BodyTable from "../common/BodyTable";

import useColWidthsResizeObserver from "./useColWidthsResizeObserver";
import Colgroup from "../common/Colgroup";
import TbodyScroller from "../common/TbodyScroller";
import ScrollContainer from "../../common/ScrollContainer";
import cx from "../../utils/cx";

const NonSticky = ({
    className,
    tbodyRef,
    scrollContainerRef,
    getRowExtraProps,
    getCellExtraProps,
    RowComponent,
    CellComponent,
    TotalsCellComponent,
    onScroll,
    ...props
}) => {

    const API = useApi();

    const { headlessMode, totals } = API;

    /*
        Hidden tfoot & thead are needed to 'hold' widths of tbody columns no to be narrower than real thead/tfoot
        and notify model about columns width changes.
        There are 3 tables rendered in this mode, so their column widths need to be synced somehow.
    */
    const widthsObserverRef = useColWidthsResizeObserver( API );

    if( process.env.NODE_ENV !== "production" ){
        if( headlessMode && !totals ){
            console.warn( "NonSticky table is rendered without headers and footers. This is not ok." )
        }
    }

    return (
        <div className={cx("afvscr-nonst",className)} {...props}>
            {headlessMode ? null : (
                <TableWrapper>
                    <Thead />
                </TableWrapper>
            )}
            <ScrollContainer ref={scrollContainerRef} onScroll={onScroll}>
                <BodyTable>
                    <Colgroup />
                    {headlessMode ? null : (
                        <Thead
                            className="afvscr-hdnwrp"
                            trRef={widthsObserverRef}
                        />
                    )}
                    {totals && (
                        <Tfoot
                            TotalsCellComponent={TotalsCellComponent}
                            className="afvscr-hdnwrp"
                            trRef={headlessMode?widthsObserverRef:undefined}
                        />
                    )}
                    <TbodyScroller />
                    <Tbody
                        tbodyRef={tbodyRef}
                        getRowExtraProps={getRowExtraProps}
                        getCellExtraProps={getCellExtraProps}
                        RowComponent={RowComponent}
                        CellComponent={CellComponent}
                    />
                </BodyTable>
            </ScrollContainer>
            {totals && (
                <TableWrapper>
                    <Tfoot TotalsCellComponent={TotalsCellComponent} />
                </TableWrapper>
            )}
        </div>
    );
};

export default observer( NonSticky );