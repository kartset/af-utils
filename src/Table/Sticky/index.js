import React, { memo, useEffect } from "react";
import Colgroup from "../common/Colgroup";
import TbodyScroller from "../common/TbodyScroller";
import ScrollContainer from "../../common/ScrollContainer";
import Thead from "../common/Thead";
import Tbody from "../common/Tbody";
import Tfoot from "../common/Tfoot";
import BodyTable from "../common/BodyTable";
import useApi from "../../useApi";
import cx from "../../utils/cx";

const SUBSCRIBE_EVENTS = [
    "#headlessMode",
    "#totals"
];


/*
    Todo:
        measure thead & tfoot heights in order to properly calculate available space for rows
*/


const Sticky = ({
    tbodyRef,
    scrollContainerRef,
    getRowExtraProps,
    getCellExtraProps,
    RowComponent,
    CellComponent,
    TotalsCellComponent,
    className,
    ...props
}) => {

    const { headlessMode, totals } = useApi( SUBSCRIBE_EVENTS );

    if( process.env.NODE_ENV !== "production" ){
        /*
            https://bugs.chromium.org/p/chromium/issues/detail?id=702927
        */

        const areTotalsPresent = totals && totals.length !== 0;

        useEffect(() => {
            if( !headlessMode || areTotalsPresent ){
                /* TODO: tests fail without this if */
                if( scrollContainerRef.current ){
                    const table = scrollContainerRef.current.querySelector( "table" );
                    const tableStyle = getComputedStyle( table );
    
                    if( tableStyle.borderCollapse === "collapse" ){
                        console.warn(
                            "You use sticky table version. Due to special border behavior when scrolling, use border-collpase: separate.%o",
                            "https://bugs.chromium.org/p/chromium/issues/detail?id=702927"
                        );
                    }
                }
            }
        }, [ headlessMode, areTotalsPresent ]);
    }
    
    return (
        <ScrollContainer ref={scrollContainerRef} className={cx("afvscr-st",className)} {...props}>
            <BodyTable>
                <Colgroup />
                {headlessMode?null:<Thead />}
                <TbodyScroller />
                <Tbody
                    tbodyRef={tbodyRef}
                    getRowExtraProps={getRowExtraProps}
                    getCellExtraProps={getCellExtraProps}
                    RowComponent={RowComponent}
                    CellComponent={CellComponent}
                />
                {totals && (
                    <Tfoot TotalsCellComponent={TotalsCellComponent} />
                )}
            </BodyTable>
        </ScrollContainer>
    );
}

export default memo( Sticky );