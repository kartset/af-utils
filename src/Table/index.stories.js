import React, { useEffect, Fragment, useState, useCallback } from "react";
import { observable, runInAction } from "mobx";
import { observer, Observer, useLocalObservable } from "mobx-react-lite";
import Table from "./index";
import r from "lodash/random";
import times from "lodash/times";
import faker from "faker";
import DefaultCellComponent from "./Cell";

export default {
    title: "Table",
    component: Table
};

export const FixedTable = () => {

    const rows = times( 90, index => ({
        a: index,
        country: faker.address.country(),
        name: faker.name.firstName(),
        height: r( 40, 200 )
    }));

    return (
        <Table
            fixedSize
            getRowData={i => rows[ i ]}
            totals={{
                a: [ "sum" ],
                country: [ "count" ]
            }}
            rows={rows}
            initialGrouping="a"
            initialExpandedGroups={{
                "1": true
            }}
            columns={[
                {
                    dataKey: "a",
                    label: "a",
                    sort: "numeric"
                },
                {
                    dataKey: "country",
                    label: "country",
                    sort: "locale"
                },
                {
                    dataKey: "name",
                    label: "name",
                    sort: "locale"
                }
            ]}
        />
    );
};

export const LiveUpdatingTable = () => {

    const CellComponent = observer(DefaultCellComponent);

    const renderHue = cellData => (
        <div
            style={{ background: `hsl(${cellData},90%,70%)` }}
        >
            hue:&nbsp;{cellData}
        </div>
    );

    const rowCount = 10000;

    const [ rows ] = useState(() => {
        const res = [];

        for( let j = 0; j < rowCount; j++ ){
            res.push({ n1: 0, n2: 0, n3: 0, n4: r( 30, 130 ) });
        }

        return observable( res );
    });

    const columns = [
        {
            dataKey: "n1",
            label: "N1",
            width: 100,
            render: renderHue
        },
        {
            dataKey: "n2",
            label: "N2",
            width: 100,
            render: renderHue
        },
        {
            dataKey: "n3",
            label: "N3",
            width: 100,
            render: renderHue
        }
    ];
    
    const getRowData = useCallback( index => rows[ index ], [ rows ]);

    useEffect(() => {
        const intervalHandle = setInterval(() => {
            runInAction(() => {
                for( let j = 0, end = rowCount / 4 | 0; j < end; j++ ){
                    for( let i = 0; i < columns.length; i++ ){
                        rows[r(0,rowCount-1)][ `n${i+1}` ] = r( 0, 360 );
                    }
                }
            });
        }, 500 );

        return () => {
            clearInterval( intervalHandle );
        };
    }, [ rows ]);

    return (
        <Table
            CellComponent={CellComponent}
            getRowData={getRowData}
            getRowExtraProps={ r => ({ style: { lineHeight: r.n4 + "px" }})}
            rows={rows}
            columns={columns}
            totals={{
                n1: [ "sum", "average", "count" ],
                n2: [ "sum", "average" ],
                n3: [ "sum", "average" ]
            }}
        />
    );
};

export const AddRemoveRows = () => {

    const getRow = useCallback( index => ({
        a: index,
        k: Math.random(),
        country: faker.address.country(),
        name: faker.name.firstName(),
        height: r( 40, 200 )
    }), []);

    const rows = useLocalObservable(() => ({
        list: times( 90, getRow ),
        get length(){
            return rows.list.length;
        },
        add(){
            for( let j = 0; j < 3000; j++ ){
                rows.list.push( getRow( j ) );
            }
        },
        remove(){
            rows.list.splice( 1, 1 );
            rows.list.pop();
        }
    }));

    const getRowData = useCallback( i => rows.list[ i ], []);

    const getRowKey = useCallback( i => rows.list[ i ].k, []);

    return (
        <Observer>
            {() => (
                <Fragment>
                    <button onClick={() => rows.add()}>Append 3000 rows</button>
                    <button onClick={() => rows.remove()}>Remove 2nd row</button>
                    <Table
                        getRowData={getRowData}
                        getRowKey={getRowKey}
                        totals={{
                            a: [ "sum" ],
                            country: [ "count" ]
                        }}
                        rows={rows}
                        columns={[
                            {
                                dataKey: "a",
                                label: "a",
                                sort: "numeric"
                            },
                            {
                                dataKey: "country",
                                label: "country",
                                sort: "locale"
                            },
                            {
                                dataKey: "name",
                                label: "name",
                                sort: "locale"
                            }
                        ]}
                    />
                </Fragment>
            )}
        </Observer>
    );
};