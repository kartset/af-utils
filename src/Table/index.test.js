import React from "react";
import { create, act } from "react-test-renderer";
import Table from "./index";

/*
    TODO: move this to config
*/
class ResizeObserver {
    observe() {
        // do nothing
    }
    unobserve() {
        // do nothing
    }
}

window.ResizeObserver = ResizeObserver;

describe( "Table component renders ok", () => {

    test.each([ undefined, -2, -1, 0, 1, 2 ].map( v => [ v ] ))( "Performs initial render with rowCount: %p", rowCount => {
        let component;
        act(() => {
            component = create(
                <Table
                    columns={[]}
                    getRowData={() => ({ a: 1 })}
                    rowCount={rowCount}
                />
            );
        });
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});