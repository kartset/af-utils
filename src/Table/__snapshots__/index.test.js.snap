// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table component renders ok Performs initial render with rowCount: -1 1`] = `null`;

exports[`Table component renders ok Performs initial render with rowCount: -2 1`] = `null`;

exports[`Table component renders ok Performs initial render with rowCount: 0 1`] = `null`;

exports[`Table component renders ok Performs initial render with rowCount: 1 1`] = `
<div
  className="afvscr-scroll-container afvscr-st afvscr-table-wrapper"
  onScroll={[Function]}
  tabIndex="0"
>
  <div
    aria-hidden="true"
    className="afvscr-height-provider"
    style={
      Object {
        "height": 16,
      }
    }
  />
  <table
    aria-colcount={0}
    aria-rowcount={1}
    className="afvscr-main-table"
  >
    <colgroup />
    <thead
      onClick={[Function]}
    >
      <tr />
    </thead>
    <tbody
      aria-hidden="true"
      className="afvscr-scroller"
    >
      <tr
        aria-hidden="true"
        className="afvscr-scroller"
        style={
          Object {
            "height": 0,
          }
        }
      />
    </tbody>
    <tbody>
      <tr
        aria-rowindex={1}
      />
    </tbody>
  </table>
</div>
`;

exports[`Table component renders ok Performs initial render with rowCount: 2 1`] = `
<div
  className="afvscr-scroll-container afvscr-st afvscr-table-wrapper"
  onScroll={[Function]}
  tabIndex="0"
>
  <div
    aria-hidden="true"
    className="afvscr-height-provider"
    style={
      Object {
        "height": 32,
      }
    }
  />
  <table
    aria-colcount={0}
    aria-rowcount={2}
    className="afvscr-main-table"
  >
    <colgroup />
    <thead
      onClick={[Function]}
    >
      <tr />
    </thead>
    <tbody
      aria-hidden="true"
      className="afvscr-scroller"
    >
      <tr
        aria-hidden="true"
        className="afvscr-scroller"
        style={
          Object {
            "height": 0,
          }
        }
      />
    </tbody>
    <tbody>
      <tr
        aria-rowindex={1}
      />
      <tr
        aria-rowindex={2}
      />
    </tbody>
  </table>
</div>
`;

exports[`Table component renders ok Performs initial render with rowCount: undefined 1`] = `null`;
